### DISCUSSION TITLE
OPNsense Support

### DISCUSSION THREAD

**Discussion by sthgi** [henrygd/beszel#39] [General]
Really love this project and its potential!  Wondering if you would consider adding support for OPNsense in Beszel?  Like Unraid they have Community Plugins.  It opens the door to networking metrics.

Thanks for creating this!

**Comment by henrygd** [comment]
Can you provide more details on what's required for a plugin? I think it's FreeBSD, right? Someone on reddit also mentioned this and wanted to test the FreeBSD binaries, so I'm going to send it to them in a bit for testing.

Edit: Also, I'm converting this to a discussion.

**Comment by sthgi** [comment]
Yes it is based off of FreeBSD 13 and 14 with their newest release.  Testing other products is what got me here maybe some of this can cut down on the time it takes you to implement something like this.  OPNsense is a offshoot of pfsense so that could be included as well.

This guy wrote a python script for OPNsense to work with Checkmk: https://github.com/bashclub/checkmk-opnsense-agent.

He was supporting:
CPU Usage
DHCP Pool
File System
Firmware Version
Gateway
Interfaces
Memory
Unbound
Uptime
WireGuard Clients

His Python script is available here:
https://github.com/bashclub/checkmk-opnsense-agent/blob/main/opnsense_checkmk_agent.py

Here is another guy using json for Zabbix:
https://grumpy.systems/2024/opnsense-monitoring-with-zabbix/

Hope this helps and obviously only promoting this for the metrics that make sense for your solution.  Hopefully others chime in as well.

Thanks again.

**Comment by sthgi** [comment]
Not finding anything to add plugins to OPNsense official repo but you can add other repos.  Here is one example:
https://www.routerperformance.net/opnsense-repo/

simple command in opnsense shell:
fetch -o /usr/local/etc/pkg/repos/mimugmail.conf https://www.routerperformance.net/mimugmail.conf

Here is the conf:
https://www.routerperformance.net/mimugmail.conf

At least with your own repo your not having to rely on anyone else.  Originally I was just thinking about an agent with this request but we could host beszel server on opnsense as well.

**Comment by utlilb** [comment]
I'm also willing to test the FreeBSD binaries. Will be on PFSense though.

**Comment by henrygd** [comment]
Here are FreeBSD amd64 binaries for testing. You can also build yourself by following the [instructions in the readme](https://github.com/henrygd/beszel/blob/main/readme.md#compiling).

```bash
GOOS=freebsd GOARCH=amd64 CGO_ENABLED=0 go build -ldflags "-w -s" .
```
[beszel-agent_freebsd_amd64.tar.gz](https://github.com/user-attachments/files/16416998/beszel-agent_freebsd_amd64.tar.gz)
[beszel_freebsd_amd64.tar.gz](https://github.com/user-attachments/files/16417012/beszel_freebsd_amd64.tar.gz)

**Comment by sthgi** [comment]
how are you guys going about auto starting the service in opnsense?

**Comment by sthgi** [comment]
For anyone else wanting to autostart bezel-agent...

90-beszel-agent -> /usr/local/etc/rc.syshook.d/start/90-beszel-agent

#!/bin/sh

echo -n "Starting Beszel Agent "
configctl beszel-agent restart

--------------------------------------------------------------------------

actions_beszel-agent.conf -> /usr/local/opnsense/service/conf/actions.d/

[start]
command:sh /usr/local/etc/rc.beszel-agent &
parameters:
type:script
message:Starting beszel-agent
description:Starting beszel-agent service

[restart]
command:sh /usr/local/etc/rc.beszel-agent &
parameters:
type:script
message:Restarting beszel-agent

[stop]
command:ps -ef | pgrep -f "beszel-agent" | xargs kill -9
parameters:
type:script
message:Stopping beszel-agent

--------------------------------------------------------------------------

beszel-agent -> /usr/local/sbin
chmod +x /usr/local/sbin/beszel-agent

--------------------------------------------------------------------------

rc.beszel-agent -> /usr/local/etc/rc.beszel-agent

#!/bin/sh

KEY="PASTE_YOUR_KEY_HERE" /usr/local/sbin/beszel-agent

---------------------------------------------------------------------------

service configd restart
configctl beszel-agent start|stop|restart
/usr/local/etc/rc.syshook.d/start/90-beszel-agent will start on reboot
actions_beszel-agent.conf will add to opnsense webui cron if it has a description under start|stop|restart

**Reply by henrygd** [reply]
Thanks for this!

Sounds like it's functional in opnsense then?

**Reply by potokslow** [reply]
Thank you very much for providing this guide, it was very useful.

I just tested this with OPNsense 24.7.12-amd64 and the official freebsd_amd64 agent binary v0.9.1.
It seems to work fine overall, except for one missing step - you forgot to mention `chmod +x /usr/local/etc/rc.syshook.d/start/90-beszel-agent` - without this the agent doesn't start on reboot.

**Reply by Cereal916** [reply]
Did you have to add a firewall rule or do something to open the port on the opnsense server? 

Edit: no need, my issue was I didn't include the first part of the key.

**Comment by sthgi** [comment]
Yes it is, its pulling the standard metrics perfectly.  I am not seeing any fluctuation with disk/io but that just could be opnsense not doing much.  I also just got it working so might just need more time monitoring to see something.  All that is left is the advanced opnsense metrics when there is time.

**Comment by duckunix** [comment]
Well, I just mainly followed this, but I built the agent myself from current code, and it working.  Not sure what it would take to create a FreeBSD ports package, nor a OPNSense package, but it is good for my needs now.

**Reply by henrygd** [reply]
We publish FreeBSD binaries now, so you should be able to update with `./beszel-agent update` if you don't want to compile again.

**Comment by tismofied** [comment]
How do I do this exactly? whenever I run the linux command from the hub I get this error
```
chown: beszel: illegal group name
Downloading and installing the agent...
Downloading and installing agent version 0.9.1 from https://github.com ...
################################################################################################################################ 100.0%
chown: beszel: illegal group name
cat: /etc/os-release: No such file or directory
Creating the systemd service for the agent...
./install-agent.sh: cannot create /etc/systemd/system/beszel-agent.service: No such file or directory

Loading and starting the agent service...
./install-agent.sh: systemctl: not found
./install-agent.sh: systemctl: not found
./install-agent.sh: systemctl: not found

Would you like to enable automatic daily updates for beszel-agent? (y/n): y
Setting up daily automatic updates for beszel-agent...
./install-agent.sh: cannot create /etc/systemd/system/beszel-agent-update.service: No such file or directory
./install-agent.sh: cannot create /etc/systemd/system/beszel-agent-update.timer: No such file or directory
./install-agent.sh: systemctl: not found
./install-agent.sh: systemctl: not found

Automatic daily updates have been enabled.
./install-agent.sh: systemctl: not found
Error: The Beszel Agent service is not running.
./install-agent.sh: systemctl: not found
```

**Reply by henrygd** [reply]
The install script doesn't support FreeBSD / OPNsense yet.

We can add support by adapting what @sthgi provided, but for now please download the binary manually and reference the comments here by stghi and potokslow to run as a service.

**Reply by mxlb-dev** [reply]
These steps worked for me, except for the OPNsense GUI entry that @sthgi mentioned. 

Download the Beszel agent binary:

```javascript
cd /usr/local/sbin
fetch https://github.com/user-attachments/files/16416998/beszel-agent_freebsd_amd64.tar.gz
```

Extract the downloaded archive:

```javascript
tar -xvzf beszel-agent_freebsd_amd64.tar.gz
```

Ensure the binary has execution permissions:

```javascript
chmod +x beszel-agent
```

Create autostart hook:

```javascript
nano /usr/local/etc/rc.syshook.d/start/90-beszel-agent
```

Content:

```javascript
#!/bin/sh

echo -n "Starting Beszel Agent "
configctl beszel-agent restart
```

Make executable:

```javascript
chmod +x /usr/local/etc/rc.syshook.d/start/90-beszel-agent
```

Create UI config:

```javascript
nano /usr/local/opnsense/service/conf/actions.d/actions_beszel-agent.conf
```

Content:

```javascript
[start]
command:sh /usr/local/etc/rc.beszel-agent &
parameters:
type:script
message:Starting beszel-agent
description:Starting beszel-agent service

[restart]
command:sh /usr/local/etc/rc.beszel-agent &
parameters:
type:script
message:Restarting beszel-agent

[stop]
command:ps -ef | pgrep -f "beszel-agent" | xargs kill -9
parameters:
type:script
message:Stopping beszel-agent
```

Create:

```javascript
nano /usr/local/etc/rc.beszel-agent
```

Content:

```javascript
#!/bin/sh

KEY="ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIJaamJxwjcK1JOwB4sMZ8ldYHSNoEB5/A+Fanz2stHrw" /usr/local/sbin/beszel-agent
```

Restart configd:

```javascript
service configd restart
```

Start / stop / restart Beszel agent:

```javascript
configctl beszel-agent start
configctl beszel-agent stop
configctl beszel-agent restart
```

**Reply by mxlb-dev** [reply]
@sthgi could you advise on where to find the gui elements you mentioned? for me there is none under system > settings > cron
or did I made a mistake in my code above. It took me several attempts and this is copy/paste from my notes. Thanks!

**Reply by tismofied** [reply]
Thank you @mxlb-dev for the step by step. I would have never figured out how to do it without them.

**Reply by mxlb-dev** [reply]
yw

**Reply by tismofied** [reply]
how do I update the agent if I followed this method?

**Reply by henrygd** [reply]
I can't test this but try running:

```bash
/usr/local/sbin/beszel-agent update
```

Then restart if it updated successfully:

```bash
configctl beszel-agent restart
```

**Reply by tismofied** [reply]
Can't believe it's this easy lol. Thanks @henrygd

**Reply by Cereal916** [reply]
> I can't test this but try running:
> 
> ```bash
> /usr/local/sbin/beszel-agent update
> ```
> 
> Then restart if it updated successfully:
> 
> ```bash
> configctl beszel-agent restart
> ```

This worked for me, but I did have to update the file permissions after I ftp the new agent bin over

chmod +x /usr/local/sbin/beszel-agent

**Comment by potokslow** [comment]
I realized I'm only seeing bandwidth information form one network interface - is this the expected behavior for FreeBSD agent, or is something broken in my setup?
Looking at the discussions about Linux agent setup, I'm guessing that Linux agent is able to access data from multiple interfaces, but is it the same for FreeBSD one?

I have 4 physical interfaces with a few virtual ones on top.
Ideally, I'd like to be able to visualize traffic on all physical and virtual interfaces, but having data from just the physical ones would be good enough.
Is it possible to collect and display data from all interfaces?

**Reply by henrygd** [reply]
Yes the behavior should be the same. The agent logs the interfaces it detects on startup.

On FreeBSD it uses the command `netstat -ibdnW` to pull the data.

Currently we combine the traffic to display in the bandwidth chart, but we'll likely move to display the interfaces separately in the future. Related issue here: #388.

Another possible cause is that the agent started on boot before all interfaces were initialized. I'll update the agent to handle this scenario at some point, but for now you should be able to avoid it by using `# REQUIRE: networking` in the rc script. There's an example in the FreeBSD guide on rc.d scripting: https://docs.freebsd.org/en/articles/rc-scripting/

Try restarting the agent and check the logs to see if it's detecting the interfaces you want.

**Comment by mxlb-dev** [comment]
Hello @henrygd queueing up for this, too. Any chance we can get an install script that supports OPNsense/FreeBSD soon? Would coffee or beer help? :)

**Reply by henrygd** [reply]
I've been working on a [public roadmap](https://github.com/users/henrygd/projects/6) today and added this. I have a lot of other stuff to do so I can't make any promises on a timeline.

**Comment by Cereal916** [comment]
I got it working, however it doesnt seem to be picking up the cpu temps, that widget is completely missing from the dashboard. It displays the cpu usage and the model correctly, but no temps. I can see the temps on the opnsense dashboard though. And they display in the terminal with "sysctl -a | grep temperature"

12th Gen Intel(R) Core(TM) i7-1255U (12c/12t)

Is this a bug, or is there something I need to configure on my end?

**Reply by henrygd** [reply]
Looks like gopsutil doesn't support reading sensors in FreeBSD.

There's an old PR for it that didn't get merged: https://github.com/shirou/gopsutil/pull/1243

I asked if they'd be interested in an updated PR.

**Reply by christopherflora** [reply]
I'm sure this hasn't been at the top of your to-do list, but I did want to ask if there's been any progress here?  I didn't see anything recent in the link above.  Just curious - temp tracking on my opnsense box is one of the only things I'm missing in beszel. 

Seriously, amazing work here, I am absolutely *loving* this tool!!  I've already found 10 things to fix in my homelab just by looking through the output.  No rush from me, but once this temp issue is resolved, and I can add my zfs volumes (https://github.com/henrygd/beszel/issues/200), then I'll be 100% covered and can retire a few poorly hacked-together bash scripts :D

**Comment by CrazyWolf13** [comment]
Create a package request at opnsense, so it hopefully gets packaged as official plugin :)
https://github.com/opnsense/plugins/issues/4579

**Comment by CrazyWolf13** [comment]
Seems like before it can get into official OPNsense, it needs to be in the freebsd-ports repository here:
https://github.com/freebsd/freebsd-ports

If anyone has the knowledge and time to open an PR over there.

**Comment by SoFMeRight** [comment]
I vibe-coded an automated installer for pfsense I know its not opnsense lol. But ehh its a solution for one of em 🤷🏽‍♀️
https://github.com/sofmeright/ant-parade-public/blob/main/cmd/monitoring/install-beszel-agent-pfsense.sh

**Reply by aldasa** [reply]
Thanks for this.

**Comment by CrazyWolf13** [comment]
Seems like we first need a FreeBSD Port, if anyone has the knowledge to do it:
https://github.com/opnsense/plugins/issues/4579

https://github.com/freebsd/freebsd-ports

**Comment by gonsalvesc** [comment]
Is anyone having trouble getting it to work with v0.12.x? All I get is 
```
2025/08/25 23:40:37 INFO Data directory path=/var/lib/beszel-agent
```

It works perfectly with v0.11.1

**Reply by chrizy66** [reply]
Updated to 0.12.4 and it stopped working, went back to 0.12.3 - same; reinstalled 0.11.1 and it is working again.

**Reply by Cereal916** [reply]
I'm having issues upgrading from 0.12.3 to 0.12.4 beszel-agent_freebsd_amd64. Redeploying my old build worked though. I had no issues upgrading my linux boxes to 0.12.4

**Reply by gonsalvesc** [reply]
Yes, Linux works just fine. Seems to be a FreeBSD issue.

**Reply by henrygd** [reply]
0.12.4 has a bug affecting at least some FreeBSD systems. Should have a fix out later today or tomorrow, until then please use 0.12.3. Sorry for the trouble!

**Reply by Cereal916** [reply]
Thanks for the update

**Reply by chrizy66** [reply]
> 0.12.4 has a bug affecting at least some FreeBSD systems. Should have a fix out later today or tomorrow, until then please use 0.12.3. Sorry for the trouble!

I tried downgrading from 0.12.4 to 0.12.3 and it didn't work on OPNsese box.

**Reply by Cereal916** [reply]
0.12.5 worked on FreeBSD with no issue.

**Comment by SoFMeRight** [comment]
I made a [script](https://raw.githubusercontent.com/sofmeright/ant-parade-public/main/cmd/monitoring/install-beszel-agent-opnsense.sh) for OpnSense. I don't currently run OpnSense as my firewall. I use pfSense but I spun up a live version of the OS to test this and try to ship a solution for those who can't run the pfSense one I made and felt left out. If you get into trouble at all and want the script to fully work for convenience we could definitely work it out together if you let me know we can make adjustments.

One liner example:
```
 fetch -o - https://raw.githubusercontent.com/sofmeright/ant-parade-public/main/cmd/monitoring/install-beszel-agent-opnsense.sh | sh -s -- -v v0.12.3 -k "ssh-ed25519 <key contents>"
```
Optional parameters:
- `-t` / `--token`
- `-u` / `--hub-url`

In my test environement I see that it successfully starts the daemon and appears to be working but my networking config is behind my pfsense firewall ~ double NAT and its having networking issues that are beyond the scope of what I want to endure to test this. I am seeing communication attempts happen but no replies. I think this is due to my environment but will need opnsense users to test and confirm it isn't working if it is to require additional adjustments.

Sample Diagnostics:
```
root@OPNsense:~ # fetch -o - https://raw.githubusercontent.com/sofmeright/ant-parade-public/main/cmd/monitoring/install-beszel-agent-opnsense.sh | sh -s -- -v v0.12.3 -k "ssh-ed25519 <redacted>"
-                                                     6382  B 5495 kBps    00s
sh: !/bin/sh: not found

Beszel Agent Installer/Updater (Provided by SoFMeRight of PrecisionPlanIT)
You can find my other promoted projects:
  GitHub: https://github.com/sofmeright
  Docker Hub: https://hub.docker.com/u/prplanit

Installing Beszel Agent version: v0.12.3 (freebsd_amd64)
Downloading https://github.com/henrygd/beszel/releases/download/v0.12.3/beszel-agent_freebsd_amd64.tar.gz ...
/tmp/beszel-agent-install-XXXXXXXX.7KdVrzzRxy/        3483 kB   19 MBps    00s
Extracting...
Installing binary to /usr/local/sbin/beszel_agent
Writing environment file at /usr/local/etc/beszel_agent.env
Creating rc.d script at /usr/local/etc/rc.d/beszel_agent
Enabling beszel_agent service...
Starting beszel_agent...
Stopping beszel_agent...
Starting beszel_agent...
Beszel Agent v0.12.3 installed and started successfully!
Manage it with: service beszel_agent {start|stop|restart|status}
Env file: /usr/local/etc/beszel_agent.env


root@OPNsense:~ # service beszel_agent status
beszel_agent is running as pid 27653
root@OPNsense:~ # tcpdump -ni vtnet0 port 45876
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on vtnet0, link-type EN10MB (Ethernet), snapshot length 262144 bytes
^[[A^C
0 packets captured
438 packets received by filter
0 packets dropped by kernel
root@OPNsense:~ # tail -n 50 /var/log/beszel_agent.log
2025/08/26 17:38:34 INFO Shutting down
2025/08/26 17:38:34 INFO Stopping SSH server
2025/08/26 17:38:34 INFO Data directory path=/var/lib/beszel-agent
2025/08/26 17:38:34 INFO Root disk mountpoint=/ io=cd0
2025/08/26 17:38:34 INFO Detected network interface name=vtnet0 sent=5874519 recv=22732385
2025/08/26 17:38:34 WARN Error creating WebSocket client err="HUB_URL environment variable not set"
2025/08/26 17:38:34 INFO Starting SSH server addr=:45876 network=tcp
root@OPNsense:~ # service beszel_agent status
beszel_agent is running as pid 27653
root@OPNsense:~ # sockstat -4 -l | grep 45876
root     beszel_age 27653 3   tcp46  *:45876               *:*
root@OPNsense:~ # sockstat -4 -l | grep 45876
root     beszel_age 27653 3   tcp46  *:45876               *:*
root@OPNsense:~ # tcpdump -ni vtnet0 port 45876
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on vtnet0, link-type EN10MB (Ethernet), snapshot length 262144 bytes
18:21:57.404461 IP 10.55.6.137.49112 > 10.55.6.234.45876: Flags [S], seq 416856667, win 64240, options [mss 1460,sackOK,TS val 502718057 ecr 0,nop,wscale 7], length 0
18:21:57.404517 IP 10.55.6.234.45876 > 10.55.6.137.49112: Flags [S.], seq 3021774159, ack 416856668, win 65228, options [mss 1460,nop,wscale 7,sackOK,TS val 1148812587 ecr 502718057], length 0
18:21:58.407095 IP 10.55.6.137.49112 > 10.55.6.234.45876: Flags [S], seq 416856667, win 64240, options [mss 1460,sackOK,TS val 502719060 ecr 0,nop,wscale 7], length 0
18:21:58.407141 IP 10.55.6.234.45876 > 10.55.6.137.49112: Flags [S.], seq 3021774159, ack 416856668, win 65228, options [mss 1460,nop,wscale 7,sackOK,TS val 1148813590 ecr 502719060], length 0
```

**Reply by henrygd** [reply]
Thank you Kai!

If you're interested in adding support directly to our install script, that's located here: https://github.com/henrygd/beszel/blob/main/supplemental/scripts/install-agent.sh

I would be happy to merge a PR for this as it would be a big help to many users :)

**Comment by chrizy66** [comment]
I updated to 0.12.5 using the `update` parameter, but the server showed version 0.11.1, so I downloaded version 0.12.5 from the repository and installed it, but the server still shows 0.11.1; `beszel-agent version` shows it correctly.
OPNsense 25.7.2

**Reply by henrygd** [reply]
Sounds like you need to restart the service. Try this:

```
configctl beszel-agent restart
```

Or if that doesn't work:

```
service beszel-agent restart
```

**Reply by mikejsutherland** [reply]
The agent reports the old version because the old agent process is still connected and running.  This always happens to me on OPNSense when updating the agent. Instead of restarting the current agent, the restart command spawns another agent process in parallel. eg:
```
root@OPNsense:~ # ps -aux | grep beszel
root    73543   0.0  0.0   14312  2644  -  I     8Aug25      0:00.00 sh /usr/local/etc/rc.beszel-agent
root    74144   0.0  0.1 1233764 14948  -  S     8Aug25      1:46.83 /usr/local/sbin/beszel-agent
root    94519   0.0  0.0   13744  2296  0  S+   23:19        0:00.00 grep beszel
root@OPNsense:~ # configctl beszel-agent restart
OK
root@OPNsense:~ # ps -aux | grep beszel
root    73543   0.0  0.0   14312  2644  -  I     8Aug25      0:00.00 sh /usr/local/etc/rc.beszel-agent
root    74144   0.0  0.1 1233764 14948  -  I     8Aug25      1:46.83 /usr/local/sbin/beszel-agent
root    95384   0.0  0.0   14312  2644  -  S    23:19        0:00.00 sh /usr/local/etc/rc.beszel-agent
root    95527   0.0  0.1 1232868 10040  -  S    23:19        0:00.01 /usr/local/sbin/beszel-agent
root    20114   0.0  0.0   13744  2284  0  S+   23:20        0:00.00 grep beszel
root@OPNsense:~ #
```

This happens so frequently I wrote a script to handle the agent update process for me. It uses pkill to stop the existing agent, I found the stop command caused some issues with my script.

beszel-agent-update.sh
```
#!/bin/sh

#configctl beszel-agent stop
echo "Shutting down beszel-agent..."
pkill -f 'beszel-agent'
ps -aux | grep -Evq 'beszel-agent$' || { echo "Failed to stop beszel-agent"; exit 1; }

beszel-agent update

echo "Starting beszel-agent..."
configctl beszel-agent start
ps -aux | grep -E 'beszel-agent$'
```

**Reply by henrygd** [reply]
Interesting, thanks. I'll see if I can set aside some time this weekend to install OPNSense and look at this.

**Reply by mikejsutherland** [reply]
@henrygd I am more than happy to help test or provide more info if needed. Great project btw!

**Reply by chrizy66** [reply]
I had to `stop` before the `update` then `start` and then it shows correct version in the server UI - thanks @mikejsutherland; there was no need to `pfkill` beszel process. I confirm that `restart` creates a new instance.
Cheers to you @henrygd  for a great work.

**Reply by mikejsutherland** [reply]
@chrizy66 correct the `stop` command works fine in the terminal, it just didn't play well inside the script I wrote so I used pkill instead.

@henrygd the two scenarios:
1) configctl beszel-agent restart -> spawns a new agent without terminating the old one
2) beszel-agent update -> (without stopping the agent prior) will start a second agent with the new version but leave the old version connected/running. Thus beszel see's the wrong version. Attempts to fix the issue by restarting the agent leads back to scenario 1).

**Comment by xbb** [comment]
If it helps, this is how I manually install, run and update beszel-agent in FreeBSD/OPNsense.

I didn't script the install since I've only two freebsd boxes.

Notes for this example:

- the agent runs as `beszel` user (I've chosen 2000 as its id, feel free to change or omit it in the command below)
- the agent can self-update, but you'll need to restart manually the service
- no DATA_DIR

Assuming the following commands are executed as root.

Create a beszel user:

```
pw user add beszel -u 2000 -d /nonexistent -s /usr/sbin/nologin -c "beszel user"
```

Create install dir location

```
mkdir -p /opt/beszel-agent/bin
```

Install beszel-agent (I've chosen an older than current version here so you can test the update):

```
curl -sL https://github.com/henrygd/beszel/releases/download/v0.12.5/beszel-agent_freebsd_amd64.tar.gz | tar -xC /opt/beszel-agent/bin beszel-agent
```

Fix permissions so that the agent can self update

```
chown -R beszel:wheel /opt/beszel-agent/bin
```

Create the env configuration file with the required permissions

```
touch /opt/beszel-agent/env && chmod 640 /opt/beszel-agent/env && chown root:beszel /opt/beszel-agent/env
```

Fill it with the required content as you need (I set the service to bind to 127.0.0.1 since I don't use SSH mode):

```
LISTEN=127.0.0.1:45876
KEY="ssh-ed25519 ..."
HUB_URL=https://example.internal
TOKEN=abcde
```

Create the rc file for the service, for convenience I place it in `/opt/beszel-agent/rc`

Give it executable permissions

```
chmod 755 /opt/beszel-agent/rc
```

Link it to the expected place

```
ln -s /opt/beszel-agent/rc /usr/local/etc/rc.d/beszel-agent
```

<details>
<summary>rc file content (click to expand)</summary>

```sh
#!/bin/sh

# PROVIDE: beszel_agent
# REQUIRE: DAEMON NETWORKING
# BEFORE: LOGIN
# KEYWORD: shutdown

# Add the following lines to /etc/rc.conf to enable Beszel Agent:
# beszel_agent_enable="YES"
#
# beszel_agent_enable (bool):   Set to YES to enable Beszel Agent
#                               Default: NO
# beszel_agent_env_file (str):  Beszel Agent env configuration file
#                               Default: /opt/beszel-agent/env
# beszel_agent_user (str):      Beszel Agent daemon user
#                               Default: beszel
# beszel_agent_bin (str):       Path to the beszel-agent binary
#                               Default: /opt/beszel-agent/bin/beszel-agent
# beszel_agent_flags (str):     Extra flags passed to beszel-agent command invocation
#                               Default:

. /etc/rc.subr

name="beszel_agent"
rcvar=beszel_agent_enable

load_rc_config $name
: ${beszel_agent_enable:="NO"}
: ${beszel_agent_user:="beszel"}
: ${beszel_agent_flags:=""}
: ${beszel_agent_env_file:="/opt/beszel-agent/env"}
: ${beszel_agent_bin:="/opt/beszel-agent/bin/beszel-agent"}

logfile="/var/log/${name}.log"
pidfile="/var/run/${name}.pid"

procname="/usr/sbin/daemon"
start_precmd="${name}_prestart"
start_cmd="${name}_start"
stop_cmd="${name}_stop"

extra_commands="upgrade"
upgrade_cmd="beszel_agent_upgrade"

beszel_agent_prestart()
{
    if [ ! -f "${beszel_agent_env_file}" ]; then
        echo WARNING: missing "${beszel_agent_env_file}" env file. Start aborted.
        exit 1
    fi
}

beszel_agent_start()
{
    echo "Starting ${name}"
    /usr/sbin/daemon -fc \
            -P "${pidfile}" \
            -o "${logfile}" \
            -u "${beszel_agent_user}" \
            "${beszel_agent_bin}" ${beszel_agent_flags}
}

beszel_agent_stop()
{
    pid="$(check_pidfile "${pidfile}" "${procname}")"
    if [ -n "${pid}" ]; then
        echo "Stopping ${name} (pid=${pid})"
        kill -- "-${pid}"
        wait_for_pids "${pid}"
    else
        echo "${name} isn't running"
    fi
}

beszel_agent_upgrade()
{
    echo "Upgrading ${name}"
    if command -v sudo >/dev/null; then
        sudo -u "${beszel_agent_user}" -- "${beszel_agent_bin}" update
    else
        su -m "${beszel_agent_user}" -c "${beszel_agent_bin} update"
    fi

    echo "Restarting ${name}"
    beszel_agent_stop
    beszel_agent_start
}

run_rc_command "$1"
```
</details>

Enable the service in `/etc/rc.conf.local` or `/etc/rc.conf`

```
beszel_agent_enable="yes"
```

Then you can use the service command to control it:

```
service beszel-agent start|stop|restart|status
```

To upgrade the agent, use sudo to execute the update command as beszel user, then restart the service as root

```
sudo -u beszel /opt/beszel-agent/bin/beszel-agent update

service beszel-agent restart
```

**EDIT**: I've added an upgrade command to the rc file, so you can just do

```
service beszel-agent upgrade
```

**Reply by sysdumped** [reply]
Just configured following your instructions, it did just works!
Thanks

**Comment by henrygd** [comment]
I added FreeBSD support to the agent install script based on @xbb's [extremely helpful comment](https://github.com/henrygd/beszel/discussions/39#discussioncomment-14264952).

If you want to update, the native command should work with service restart after the next release. At which point the auto-updates should also work, hopefully.

```bash
/usr/local/sbin/beszel-agent update
```

Edit: To run the install script, just copy the Linux install command in the hub.

Update: I made some updates to the FreeBSD install to better support OPNsense, the binary is now located at `/usr/local/sbin/beszel-agent` since OPNsense doesn't have `/opt` by default.

**Reply by CrazyWolf13** [reply]
@henrygd hi henry, 
Would it nevertheless be possible to get a more "official" port for freebsd, now that all the installer commands are there for freebsd?

https://ports.freebsd.org/cgi/ports.cgi?query=beszel&stype=all&sektion=all
https://docs.freebsd.org/en/books/handbook/ports/

Thanks!

**Reply by henrygd** [reply]
I looked into it, but I don't have time to figure it out at the moment.

**Comment by CrazyWolf13** [comment]
Failure to start beszel service after reboot.

@henrygd thanks a lot for that easy installer via command, it was really easy to install beszel with the command generated from the webui.
That works wonderfully, e.g. the hub shows stats, until I reboot, after a reboot beszel does not seem to send any data back.

The service does seem to start, but it complains for a missing pubkey:

```bash
root@OPNsense:/home/tobias # /usr/local/sbin/beszel-agent update
Fetching release information...
You already have the latest version 0.12.7.
root@OPNsense:/home/tobias # /usr/local/sbin/beszel-agent status
2025/09/12 21:00:16 Failed to load public keys:no key provided: must set -key flag, KEY env var, or KEY_FILE env var. Use 'beszel-agent help' for usage
root@OPNsense:/home/tobias #
```

When I then re-run the initial install command, it does work again till the next reboot.

Any ideas what might be wrong here?

**Reply by henrygd** [reply]
I haven't been able to reproduce this, unfortunately. Does the env file still exist with your content after rebooting?

```bash
cat /usr/local/etc/beszel-agent/env
```

Here's the service content if someone knows more than me about how OPNsense works.

https://github.com/henrygd/beszel/blob/8da1ded73ea6dd342cc2924f135b5e0c81283f88/supplemental/scripts/install-agent.sh#L77-L160

**Reply by CrazyWolf13** [reply]
@henrygd The problem seems to be that the beszel user is not created. 

```bash
root@OPNsense:/home/tobias # sysrc -a | grep beszel
beszel_agent_enable: YES
daemon: unknown user: beszel
daemon: unknown user: beszel
```

Compiling bash from ports repository then manually running the commands inside bash to generate the user from the script worked for me.

**Reply by henrygd** [reply]
Thanks, I think this was fixed in the revisions I made I few days ago. My guess is a UID conflict with an existing user, but I removed the fixed ID on user creation.
